lib/transport/transport-udp-socket.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2002-2019 Balabit
+ * Copyright (c) 1998-2019 Balázs Scheidler
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "transport-udp-socket.h"
+#include "transport/transport-socket.h"
+#include "scratch-buffers.h"
+#include "str-format.h"
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <errno.h>
+
+typedef struct _LogTransportUDP LogTransportUDP;
+struct _LogTransportUDP
+{
+  LogTransportSocket super;
+  GSockAddr *local_addr;
+};
+
+#if defined(SYSLOG_NG_HAVE_CTRLBUF_IN_MSGHDR)
+static void
+_feed_aux_from_cmsg(LogTransportUDP *self, LogTransportAuxData *aux, struct msghdr *msg)
+{
+  struct cmsghdr *cmsg;
+
+  for (cmsg = CMSG_FIRSTHDR(msg); cmsg != NULL; cmsg = CMSG_NXTHDR(msg, cmsg))
+    {
+
+#ifdef __FreeBSD__
+      if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_RECVDSTADDR)
+        {
+          gint cmsg_header_len = (gchar *) CMSG_DATA(cmsg) - (gchar *) cmsg;
+          struct sockaddr *sa = (struct sockaddr *) CMSG_DATA(cmsg);
+
+          log_transport_aux_data_set_local_addr_ref(aux, g_sockaddr_new(sa, cmsg->cmsg_len - cmsg_header_len));
+          break;
+        }
+#else
+      if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_PKTINFO)
+        {
+          struct sockaddr_in sin;
+          struct in_pktinfo *inpkt = (struct in_pktinfo *) CMSG_DATA(cmsg);
+
+          g_assert(self->local_addr->sa.sa_family == AF_INET);
+          sin = *(struct sockaddr_in *) &self->local_addr->sa;
+          sin.sin_addr = inpkt->ipi_addr;
+          log_transport_aux_data_set_local_addr_ref(aux, g_sockaddr_new((struct sockaddr *) &sin, sizeof(sin)));
+          break;
+        }
+#endif
+      if (cmsg->cmsg_level == IPPROTO_IPV6 && cmsg->cmsg_type == IPV6_PKTINFO)
```

Comment:
`SYSLOG_NG_ENABLE_IPV6` is missing here?

Reply:
done in 383531132